lib/fuser_mqubit.h

```diff
@@ -0,0 +1,1087 @@
+
+    while (may_have_gates_to_fuse) {
+      may_have_gates_to_fuse = false;
+
+      std::sort(links.begin(), links.end(),
+                [](const Link* l, const Link* r) -> bool {
+                  auto lp = l->prev;
+                  auto rp = r->prev;
+
+                  if (lp != nullptr && rp != nullptr) {
+                    return lp->val->parent->time > rp->val->parent->time;
+                  } else {
+                    return lp != nullptr || rp == nullptr;
+                  }
+                });
+
+      for (auto link : links) {
+        if (link->prev == nullptr) continue;
+
+        auto p = link->prev->val;
+
+        if (!QubitsAreIn(fgate.mask, p->mask) || (p->mask & bad_mask) != 0
+            || p->visited > level || p->parent->unfusible) {
+          bad_mask |= p->mask;
+        } else {
+          AddGates(level, *p, gates);
+
+          may_have_gates_to_fuse = true;
+          break;
+        }
+      }
+    }
+
+    for (auto it = gates.rbegin(); it != gates.rend(); ++it) {
+      fgate.gates.push_back(*it);
+    }
+  }
+
+  // Fuse smaller gates with fgate forward in gate time.
+  static void FuseNext(unsigned level, GateF& fgate) {
+    uint64_t bad_mask = 0;
+    auto links = fgate.links;
+
+    bool may_have_gates_to_fuse = true;
+
+    while (may_have_gates_to_fuse) {
+      may_have_gates_to_fuse = false;
+
+      auto time = fgate.parent->time;
+
+      std::sort(links.begin(), links.end(),
+                [&time](const Link* l, const Link* r) -> bool {
+                  auto ln = l->next;
+                  auto rn = r->next;
+
+                  if (ln != nullptr && rn != nullptr) {
+                    return ln->val->parent->time < rn->val->parent->time;
+                  } else {
+                    return ln != nullptr || rn == nullptr;
+                  }
+                });
+
+      for (auto link : links) {
+        if (link->next == nullptr) continue;
+
+        auto n = link->next->val;
+
+        if (!QubitsAreIn(fgate.mask, n->mask) || (n->mask & bad_mask) != 0
+            || n->visited > level || n->parent->unfusible) {
+          bad_mask |= n->mask;
+        } else {
+          AddGates(level, *n, fgate.gates);
+
+          may_have_gates_to_fuse = true;
+          break;
+        }
+      }
+    }
+  }
+
+  static void AddGates(
+      unsigned level, GateF& fgate, std::vector<const Gate*>& gates) {
+    fgate.visited = level == 0 ? 1 : 999;
+
+    if (level == 0) {
+      gates.push_back(fgate.parent);
+    } else {
+      for (auto gate :fgate.gates) {
+        gates.push_back(gate);
+      }
+    }
+
+    for (auto link : fgate.links) {
+      LinkManager::Delete(link);
+    }
+  }
+
+  static bool QubitsAreIn(uint64_t mask0, uint64_t mask) {
+    return ((mask0 | mask) ^ mask0) == 0;
+  }
+
+  static void PrintStat(unsigned verbosity, const Stat& stat,
+                        const std::vector<GateFused>& fused_gates) {
+    if (verbosity > 0) {
```

Comment:
To reduce indentation, I recommend `if (verbosity <= 0) return;` instead.

Reply:
Done.